src/REPL/replxx/Util.cpp

```diff
@@ -0,0 +1,133 @@
+#include <Ark/REPL/replxx/Util.hpp>
+#include <cstring>
+#include <regex>
+
+int utf8str_codepoint_len(char const *s, int utf8len) 
+{
+    int codepointLen = 0;
+    unsigned char m4 = 128 + 64 + 32 + 16;
+    unsigned char m3 = 128 + 64 + 32;
+    unsigned char m2 = 128 + 64;
+
+    for(int i = 0; i < utf8len; ++ i, ++ codepointLen)
+    {
+        char c = s[i];
+
+        if ((c & m4) == m4)
+        {
+            i += 3;
+        }
+        else if((c & m3) == m3)
+        {
+            i += 2;
+        }
+        else if(( c & m2 ) == m2)
+        {
+            i += 1;
+        }
+    }
+
+    return codepointLen;
+}
+
+int context_len(char const *prefix)
+{
+    char const wb[] = " \t\n\r\v\f-=+*&^%$#@!,./?<>;:`~'\"[]{}()\\|";
+    int i = std::strlen(prefix) - 1;
+    int cl = 0;
+
+    while(i >= 0)
+    {
+        if(std::strchr(wb, prefix[i]) != NULL)
+        {
+            break;
+        }
+
+        ++ cl;
+        -- i;
+    }
+
+    return cl;
+}
+
+Replxx::completions_t hook_completion(std::string const& context, int& contextLen, std::vector<std::string> const& examples)
+{
+    Replxx::completions_t completions;
+    int utf8ContextLen = context_len(context.c_str());
+    int prefixLen = context.length() - utf8ContextLen;
+
+    if((prefixLen > 0) && (context[prefixLen - 1] == '\\'))
+    {
+        -- prefixLen;
+        ++ utf8ContextLen;
+    }
+
+    contextLen = utf8str_codepoint_len(context.c_str() + prefixLen, utf8ContextLen);
+
+    std::string prefix = context.substr(prefixLen);
+    for(auto const& e : examples)
+    {
+        if(e.compare(0, prefix.size(), prefix) == 0)
```

Comment:
maybe add a comment to tell what this does (checking for a prefix at the beginning of a string) and why